unix/x0vncserver/XDesktop.cxx

```diff
@@ -331,6 +355,217 @@ void XDesktop::keyEvent(rdr::U32 keysym, rdr::U32 xtcode, bool down) {
 void XDesktop::clientCutText(const char* str, int len) {
 }
 
+ScreenSet XDesktop::computeScreenLayout()
+{
+  ScreenSet layout;
+
+#ifdef HAVE_XRANDR
+  XRRScreenResources *res = XRRGetScreenResources(dpy, DefaultRootWindow(dpy));
+  if (!res) {
+    vlog.error("XRRGetScreenResources failed");
+    return layout;
+  }
+  vncSetGlueContext(dpy, res);
+
+  layout = ::computeScreenLayout(DefaultScreen(dpy), &outputIdMap,
+				 pb->width(), pb->height());
+  XRRFreeScreenResources(res);
+#endif
+
+  return layout;
+}
+
+#ifdef HAVE_XRANDR
+/* Get the biggest mode which is equal or smaller to requested
+   size. If no such mode exists, return the smallest. */
+void GetSmallerMode(XRRScreenResources *res,
+		    XRROutputInfo *output,
+		    unsigned int *width, unsigned int *height)
+{
+  XRRModeInfo best = {};
+  XRRModeInfo smallest = {};
+  smallest.width = -1;
+  smallest.height = -1;
+
+  for (int i = 0; i < res->nmode; i++) {
+    for (int j = 0; j < output->nmode; j++) {
+      if (output->modes[j] == res->modes[i].id) {
+	if ((res->modes[i].width > best.width && res->modes[i].width <= *width) &&
+	    (res->modes[i].height > best.height && res->modes[i].height <= *height)) {
+	  best = res->modes[i];
+	}
+	if ((res->modes[i].width < smallest.width) && res->modes[i].height < smallest.height) {
+	  smallest = res->modes[i];
+	}
+      }
+    }
+  }
+
+  if (best.id == 0 && smallest.id != 0) {
+    best = smallest;
+  }
+
+  *width = best.width;
+  *height = best.height;
+}
+#endif /* HAVE_XRANDR */
+
+unsigned int XDesktop::setScreenLayout(int fb_width, int fb_height,
+				       const rfb::ScreenSet& layout)
+{
+#ifdef HAVE_XRANDR
+  char buffer[2048];
+  vlog.debug("Got request for framebuffer resize to %dx%d",
+             fb_width, fb_height);
+  layout.print(buffer, sizeof(buffer));
+  vlog.debug("%s", buffer);
+
+  XRRScreenResources *res = XRRGetScreenResources(dpy, DefaultRootWindow(dpy));
+  if (!res) {
+    vlog.error("XRRGetScreenResources failed");
+    return rfb::resultProhibited;
+  }
+  vncSetGlueContext(dpy, res);
+
+  /* The client may request a screen layout which is not supported by
+     the Xserver. This happens, for example, when adjusting the size
+     of a non-fullscreen vncviewer window. To handle this and other
+     cases, we first call setScreenLayout with tryrun=true. If this
+     fails, we try to adjust the request to one screen with a smaller
+     mode. */
+  OutputIdMap dryrunIdMap = outputIdMap;
+  vlog.debug("Testing screen layout");
+  unsigned int tryresult = ::setScreenLayout(true, 0, vlog, fb_width, fb_height,
```

Comment:
Bike shedding, but perhaps having a separate function called `tryScreenLayout()` is more readable?